src/connect.rs

```diff
@@ -960,6 +977,21 @@ mod socks {
     }
 }
 
+#[cfg(unix)]
+mod unix_socket_conn {
+    use std::os::unix::io::OwnedFd;
+    use std::path::Path;
+    use tokio::net::{TcpStream, UnixStream};
+    use crate::error::BoxError;
+
+    pub async fn connect<P: AsRef<Path>>(socket: P) -> Result<TcpStream, BoxError> {
+        let target_stream = UnixStream::connect(&socket).await?;
+        let owned_fd: OwnedFd = target_stream.into_std()?.into();
+        let stream = std::net::TcpStream::from(owned_fd);
```

Comment:
This is perfectly fine? To simply cast a Unix Domain Stream into a TCP stream? If that's the case, it'd be useful to include in the comments a link to the proof of that.

Reply:
To argue that it works to make TcpStream use a file descriptor created by UnixStream, we can split the usage in 2 scenarios:

1. What happens if TcpStream is used for protocol-agnostic things like reading and writing? In this scenario we have to make sure, that our connection behaves correctly as a HTTP-over-UDS.
2. What happens, if TcpStream assumes the fd to have the `AF_INET` or `AF_INET6` `af_family` and tries to use protocol-specific operations like calling [setsockopt](https://pubs.opengroup.org/onlinepubs/9699919799/functions/setsockopt.html) with `level = IPPROTO_IP`? This would happend for example if `std::net::tcp::TcpStream::set_ttl` would be called. In this scenario, our HTTP-over-UDS doesn't have any sensible Ok behavior (in the example, because unix socket connections don't have a TTL), so we only have to make sure, we give a reasonable error message and don't crash.

Reasoning for the first scenario:
Comparing the common fn's between `tokio::net::unix::stream::UnixStream` and `tokio::net::tcp::stream::TcpStream` (f.e. `ready`, `readable`, `poll_read_ready`, `try_read`, `try_read_vectored`, `try_read_buf`, `writable`, `poll_write_ready`, ...), they are exactly the same. They use the same code to forward to `mio::net::uds::stream` and `mio::net::TcpStream` respectively, who's `std::io::Read` and `std::io::Write` implementations are also the same and forward to `std::net`. Both `std::net::UnixStream` and `std::net::TcpStream` forward their protocol-agnostic calls to `std::sys::unix::net::Socket`, so the common operations always result in the same system calls, regardless of whether we issue them through `tokio::net::unix::stream::UnixStream` or `tokio::net::tcp::stream::TcpStream`. It doesn't matter which of them we use. This also won't change in the future, because the fact that all sockets are used the same¹, is build into the POSIX system interfaces specification. The specifications of [recv](https://pubs.opengroup.org/onlinepubs/9699919799/functions/recv.html), [send](https://pubs.opengroup.org/onlinepubs/9699919799/functions/send.html), [read](https://pubs.opengroup.org/onlinepubs/9699919799/functions/read.html), etc. don't define any `af_family` dependent behavior.

Reasoning for the second scenario:
On unix, the calls on TcpStream get forwarded to `std::sys::unix::net`. All OS-Calls there are wrapped in `io::Result`. A non-sense `IPPROTO_IP` specific call like one to `setsockopt`, would result in a `ENOPROTOOPT` OS-error, which translates to rust's `io::Result::Err` with `ErrorKind::Uncategorized`. Similarily trying to use an operation, that relies on a a INET specific sockaddr layout, (for example if we try to reconnect to another IP addr with a new connect syscall), the OS would return a `EAFNOSUPPORT`, which again translates to `io::Result::Err` with `ErrorKind::Uncategorized` in rust. This behavior isn't particulary helpful, because the error message has to be analyzed with `io::Error::raw_os_error` and the libc-crate, to get a useful error message. But considering a `IPPROTO_IP` specific operation can't be done by a HTTP-over-UDS connection, returning `io::Error` is the intended behavior.

¹ Exept for the sockaddr variant in [connect](https://pubs.opengroup.org/onlinepubs/9699919799/functions/connect.html) calls, which raises a `EAFNOSUPPORT` error, if it doesn't match the `af_family`. But we only reinterpret the fd after the `connect` call.